File: backend/src/main/java/com/soccerdreamfermana/service/CalendarInviteService.java
package com.soccerdreamfermana.service;

import com.soccerdreamfermana.model.Event;
import java.nio.charset.StandardCharsets;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.Locale;
import org.springframework.stereotype.Service;

@Service
public class CalendarInviteService {

 private static final ZoneId EVENT_ZONE = ZoneId.of("Europe/Rome");
 private static final DateTimeFormatter ICS_FORMATTER = DateTimeFormatter.ofPattern("yyyyMMdd'T'HHmmss'Z'", Locale.ITALY);

 public byte[] buildEventInvite(Event event) {
  LocalDateTime endDate = event.getDataFine() != null ? event.getDataFine() : event.getData().plusHours(2);

  String ics = String.join("\r\n",
   "BEGIN:VCALENDAR",
   "VERSION:2.0",
   "PRODID:-//Associazione Disabili//Eventi//IT",
   "CALSCALE:GREGORIAN",
   "METHOD:PUBLISH",
   "BEGIN:VEVENT",
   "UID:event-" + event.getId() + "@soccerdreamfermana.local",
   "DTSTAMP:" + toUtc(LocalDateTime.now()),
   "DTSTART:" + toUtc(event.getData()),
   "DTEND:" + toUtc(endDate),
   "SUMMARY:" + escape(event.getTitolo()),
   "DESCRIPTION:" + escape(event.getDescrizione()),
   "LOCATION:" + escape(event.getLuogo()),
   "END:VEVENT",
   "END:VCALENDAR",
   ""
  );

  return ics.getBytes(StandardCharsets.UTF_8);
 }

 public String buildFileName(Event event) {
  String slug = event.getTitolo().toLowerCase(Locale.ITALY).replaceAll("[^a-z0-9]+", "-").replaceAll("(^-|-$)", "");
  return (slug.isBlank() ? "evento" : slug) + ".ics";
 }

 private String toUtc(LocalDateTime dateTime) {
  return dateTime.atZone(EVENT_ZONE).withZoneSameInstant(ZoneOffset.UTC).format(ICS_FORMATTER);
 }

 private String escape(String value) {
  if (value == null) {
   return "";
  }
  return value
   .replace("\\", "\\\\")
   .replace(",", "\\,")
   .replace(";", "\\;")
   .replace("\n", "\\n");
 }
}

File: backend/Services/CalendarInviteService.cs
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;
using SoccerDreamFermana.Backend.Models;

namespace SoccerDreamFermana.Backend.Services;

public sealed partial class CalendarInviteService
{
    private static readonly TimeZoneInfo EventZone = ResolveRomeTimeZone();
    private static readonly CultureInfo ItalianCulture = CultureInfo.GetCultureInfo("it-IT");

    public byte[] BuildEventInvite(Event item)
    {
        var endDate = item.DataFine ?? item.Data.AddHours(2);
        var ics = string.Join("\r\n",
            "BEGIN:VCALENDAR",
            "VERSION:2.0",
            "PRODID:-//Associazione Disabili//Eventi//IT",
            "CALSCALE:GREGORIAN",
            "METHOD:PUBLISH",
            "BEGIN:VEVENT",
            $"UID:event-{item.Id}@soccerdreamfermana.local",
            $"DTSTAMP:{ToUtc(DateTime.Now)}",
            $"DTSTART:{ToUtc(item.Data)}",
            $"DTEND:{ToUtc(endDate)}",
            $"SUMMARY:{Escape(item.Titolo)}",
            $"DESCRIPTION:{Escape(item.Descrizione)}",
            $"LOCATION:{Escape(item.Luogo)}",
            "END:VEVENT",
            "END:VCALENDAR",
            ""
        );

        return Encoding.UTF8.GetBytes(ics);
    }

    public string BuildFileName(Event item)
    {
        var slug = NonSlugRegex().Replace(item.Titolo.ToLower(ItalianCulture), "-").Trim('-');
        return $"{(string.IsNullOrWhiteSpace(slug) ? "evento" : slug)}.ics";
    }

    private static string ToUtc(DateTime dateTime)
    {
        var unspecified = DateTime.SpecifyKind(dateTime, DateTimeKind.Unspecified);
        var utc = TimeZoneInfo.ConvertTimeToUtc(unspecified, EventZone);
        return utc.ToString("yyyyMMdd'T'HHmmss'Z'", CultureInfo.InvariantCulture);
    }

    private static string Escape(string? value)
    {
        return (value ?? "")
            .Replace("\\", "\\\\", StringComparison.Ordinal)
            .Replace(",", "\\,", StringComparison.Ordinal)
            .Replace(";", "\\;", StringComparison.Ordinal)
            .Replace("\n", "\\n", StringComparison.Ordinal);
    }

    private static TimeZoneInfo ResolveRomeTimeZone()
    {
        try
        {
            return TimeZoneInfo.FindSystemTimeZoneById("Europe/Rome");
        }
        catch (TimeZoneNotFoundException)
        {
            return TimeZoneInfo.FindSystemTimeZoneById("W. Europe Standard Time");
        }
    }

    [GeneratedRegex("[^a-z0-9]+")]
    private static partial Regex NonSlugRegex();
}
